// braintree/lib/braintree/braintree_gateway.js
// Generated by CoffeeScript 1.6.1
var AddressGateway, BraintreeGateway, CreditCardGateway, CreditCardVerificationGateway, CustomerGateway, DisbursementGateway, Http, MerchantAccountGateway, SettlementBatchSummaryGateway, SubscriptionGateway, TransactionGateway, TransparentRedirectGateway, WebhookNotificationGateway, WebhookTestingGateway;

Http = require('./http').Http;

AddressGateway = require("./address_gateway").AddressGateway;

CreditCardGateway = require("./credit_card_gateway").CreditCardGateway;

CreditCardVerificationGateway = require("./credit_card_verification_gateway").CreditCardVerificationGateway;

CustomerGateway = require("./customer_gateway").CustomerGateway;

MerchantAccountGateway = require("./merchant_account_gateway").MerchantAccountGateway;

SettlementBatchSummaryGateway = require("./settlement_batch_summary_gateway").SettlementBatchSummaryGateway;

SubscriptionGateway = require("./subscription_gateway").SubscriptionGateway;

TransactionGateway = require("./transaction_gateway").TransactionGateway;

DisbursementGateway = require("./disbursement_gateway").DisbursementGateway;

TransparentRedirectGateway = require("./transparent_redirect_gateway").TransparentRedirectGateway;

WebhookNotificationGateway = require("./webhook_notification_gateway").WebhookNotificationGateway;

WebhookTestingGateway = require("./webhook_testing_gateway").WebhookTestingGateway;

BraintreeGateway = (function() {

  function BraintreeGateway(config) {
    this.config = config;
    this.http = new Http(this.config);
    this.address = new AddressGateway(this);
    this.creditCard = new CreditCardGateway(this);
    this.creditCardVerification = new CreditCardVerificationGateway(this);
    this.customer = new CustomerGateway(this);
    this.disbursement = new DisbursementGateway(this);
    this.merchantAccount = new MerchantAccountGateway(this);
    this.settlementBatchSummary = new SettlementBatchSummaryGateway(this);
    this.subscription = new SubscriptionGateway(this);
    this.transaction = new TransactionGateway(this);
    this.transparentRedirect = new TransparentRedirectGateway(this);
    this.webhookNotification = new WebhookNotificationGateway(this);
    this.webhookTesting = new WebhookTestingGateway(this);
  }

  return BraintreeGateway;

})();

exports.BraintreeGateway = BraintreeGateway;
