// braintree/lib/braintree/webhook_notification_gateway.js
// Generated by CoffeeScript 1.6.1
var Digest, Gateway, InvalidSignatureError, Util, WebhookNotification, WebhookNotificationGateway, xml2js,
  __hasProp = {}.hasOwnProperty,
  __extends = function(child, parent) { for (var key in parent) { if (__hasProp.call(parent, key)) child[key] = parent[key]; } function ctor() { this.constructor = child; } ctor.prototype = parent.prototype; child.prototype = new ctor(); child.__super__ = parent.prototype; return child; };

xml2js = require('xml2js');

Digest = require('./digest').Digest;

Gateway = require('./gateway').Gateway;

InvalidSignatureError = require('./exceptions').InvalidSignatureError;

Util = require('./util').Util;

WebhookNotification = require('./webhook_notification').WebhookNotification;

WebhookNotificationGateway = (function(_super) {

  __extends(WebhookNotificationGateway, _super);

  function WebhookNotificationGateway(gateway) {
    this.gateway = gateway;
    this.parser = new xml2js.Parser({
      explicitRoot: true
    });
  }

  WebhookNotificationGateway.prototype.parse = function(signature, payload, callback) {
    var xmlPayload,
      _this = this;
    if (!this.validateSignature(signature, payload)) {
      callback(InvalidSignatureError(), null);
      return;
    }
    xmlPayload = new Buffer(payload, "base64").toString("utf8");
    return this.parser.parseString(xmlPayload, function(err, result) {
      var attributes, handler;
      attributes = Util.convertNodeToObject(result);
      handler = _this.createResponseHandler("notification", WebhookNotification, function(err, result) {
        return callback(null, result.notification);
      });
      return handler(null, attributes);
    });
  };

  WebhookNotificationGateway.prototype.validateSignature = function(signature, payload) {
    var pair, signaturePairs;
    signaturePairs = (function() {
      var _i, _len, _ref, _results;
      _ref = signature.split("&");
      _results = [];
      for (_i = 0, _len = _ref.length; _i < _len; _i++) {
        pair = _ref[_i];
        if (pair.indexOf("|") !== -1) {
          _results.push(pair.split("|"));
        }
      }
      return _results;
    })();
    return Digest.secureCompare(this.matchingSignature(signaturePairs), Digest.hexdigest(this.gateway.config.privateKey, payload));
  };

  WebhookNotificationGateway.prototype.verify = function(challenge) {
    var digest;
    digest = Digest.hexdigest(this.gateway.config.privateKey, challenge);
    return "" + this.gateway.config.publicKey + "|" + digest;
  };

  WebhookNotificationGateway.prototype.matchingSignature = function(signaturePairs) {
    var publicKey, signature, _i, _len, _ref;
    for (_i = 0, _len = signaturePairs.length; _i < _len; _i++) {
      _ref = signaturePairs[_i], publicKey = _ref[0], signature = _ref[1];
      if (this.gateway.config.publicKey === publicKey) {
        return signature;
      }
    }
    return null;
  };

  return WebhookNotificationGateway;

})(Gateway);

exports.WebhookNotificationGateway = WebhookNotificationGateway;
