// braintree/lib/braintree.js
// Generated by CoffeeScript 1.6.1
var BraintreeGateway, Config, CreditCard, Environment, Subscription, Transaction, WebhookNotification, connect, errorTypes;

Config = require("./braintree/config").Config;

Environment = require("./braintree/environment").Environment;

BraintreeGateway = require("./braintree/braintree_gateway").BraintreeGateway;

errorTypes = require("./braintree/error_types");

Transaction = require("./braintree/transaction").Transaction;

CreditCard = require("./braintree/credit_card").CreditCard;

Subscription = require("./braintree/subscription").Subscription;

WebhookNotification = require("./braintree/webhook_notification").WebhookNotification;

connect = function(config) {
  return new BraintreeGateway(new Config(config));
};

exports.connect = connect;

exports.version = '1.14.0';

exports.Environment = Environment;

exports.errorTypes = errorTypes;

exports.Transaction = Transaction;

exports.CreditCard = CreditCard;

exports.Subscription = Subscription;

exports.WebhookNotification = WebhookNotification;
